mw_jwt_test.go

```diff
@@ -86,39 +87,84 @@ func createJWTSessionWithRSAWithPolicy(policyID string) *user.SessionState {
 	return session
 }
 
-// JWTSessionHMAC
+type JwtCreator func() *user.SessionState
+
+var createHmacJWKSession = func() *user.SessionState { return createJWTSession() }
+var createRsaJWKSession = func() *user.SessionState { return createJWTSessionWithRSA() }
+
+func prepareGenericJWTSession(testName string, method string, claimName string, GlobalSkipKidAsId bool, ApiSkipKidAsId bool) (*APISpec, string) {
+
+	tokenKID := testKey(testName, "token")
+
+	globalConf := config.Global()
+	globalConf.JWTSkipCheckKidAsId = GlobalSkipKidAsId
+	config.SetGlobal(globalConf)
 
-func prepareJWTSessionHMAC(tb testing.TB) string {
 	spec := buildAndLoadAPI(func(spec *APISpec) {
 		spec.UseKeylessAccess = false
-		spec.JWTSigningMethod = "hmac"
+		spec.JWTSigningMethod = method
 		spec.EnableJWT = true
 		spec.Proxy.ListenPath = "/"
+		spec.JWTSkipCheckKidAsId = ApiSkipKidAsId
+
+		if claimName != KID {
+			spec.JWTIdentityBaseField = claimName
+		}
+
 	})[0]
 
-	session := createJWTSession()
-	tokenKID := testKey(tb, "token")
-	spec.SessionManager.UpdateSession(tokenKID, session, 60, false)
+	var jwtToken string
+	var sessionFunc JwtCreator
+	switch method {
+	case HMACSign:
+		sessionFunc = createHmacJWKSession
+
+		jwtToken = createJWKTokenHMAC(func(t *jwt.Token) {
+			t.Claims.(jwt.MapClaims)["foo"] = "bar"
+			t.Claims.(jwt.MapClaims)["exp"] = time.Now().Add(time.Hour * 72).Unix()
+
+			if claimName != KID {
+				t.Claims.(jwt.MapClaims)[claimName] = tokenKID
+				t.Header[KID] = "ignore-this-id"
+			} else {
+				t.Header[KID] = tokenKID
+			}
+		})
+	case RSASign:
+		sessionFunc = createRsaJWKSession
+
+		jwtToken = createJWKToken(func(t *jwt.Token) {
+			t.Claims.(jwt.MapClaims)["foo"] = "bar"
+			t.Claims.(jwt.MapClaims)["exp"] = time.Now().Add(time.Hour * 72).Unix()
+
+			if claimName != KID {
+				t.Claims.(jwt.MapClaims)[claimName] = tokenKID
+				t.Header[KID] = "ignore-this-id"
+			} else {
+				t.Header[KID] = tokenKID
+			}
+		})
+	default:
+		panic("Bad signing method")
```

Comment:
don't know specifics of implementation, but should we panic here? any better way of gracefully dealing with issue?

Reply:
We can but I'll have to add checks in all the tests to verify `spec` and `token`, since there weren't any before and I haven't added it myself.